src/main/kotlin/org/rust/ide/annotator/RsMacroExpansionHighlightingPass.kt

```diff
@@ -0,0 +1,153 @@
+import org.rust.stdext.removeLast
+
+class RsMacroExpansionHighlightingPassFactory(
+    val project: Project,
+    registrar: TextEditorHighlightingPassRegistrar
+) : DirtyScopeTrackingHighlightingPassFactory {
+    private val myPassId: Int = registrar.registerTextEditorHighlightingPass(
+        this,
+        null,
+        null,
+        false,
+        -1
+    )
+
+    override fun createHighlightingPass(file: PsiFile, editor: Editor): TextEditorHighlightingPass? {
+        if (project.macroExpansionManager.macroExpansionMode !is MacroExpansionMode.New) return null
+        val restrictedRange = FileStatusMap.getDirtyTextRange(editor, passId) ?: return null
+        return RsMacroExpansionHighlightingPass(file, restrictedRange, editor.document)
+    }
+
+    override fun getPassId(): Int = myPassId
+}
+
+class RsMacroExpansionHighlightingPass(
+    private val file: PsiFile,
+    private val restrictedRange: TextRange,
+    document: Document
+) : TextEditorHighlightingPass(file.project, document) {
+    private val results = mutableListOf<HighlightInfo>()
+
+    private fun createAnnotators(): List<Annotator> = listOf(
+        RsEdition2018KeywordsAnnotator(),
+        RsHighlightingAnnotator(),
+        RsHighlightingMutableAnnotator()
+    )
+
+    override fun doCollectInformation(progress: ProgressIndicator) {
+        val macros = mutableListOf<PreparedMacroCall>()
+        for (macro in file.descendantsOfType<RsMacroCall>()) {
+            if (macro.macroArgument?.textRange?.intersects(restrictedRange) != true) continue
+            macros += macro.prepare() ?: continue
+        }
+        if (macros.isEmpty()) return
+
+        val annotators = createAnnotators()
+        val holder = AnnotationHolderImpl(AnnotationSession(file))
+        while (macros.isNotEmpty()) {
+            val macro = macros.removeLast()
+            for (element in macro.elementsForHighlighting) {
+                for (ann in annotators) {
+                    ProgressManager.checkCanceled()
+                    ann.annotate(element, holder)
+                }
+                if (element is RsMacroCall) {
+                    macros += element.prepare() ?: continue
+                }
+            }
+            mapAndCollectAnnotations(macro, holder)
+            holder.clear()
+        }
+    }
+
+    private fun mapAndCollectAnnotations(macro: PreparedMacroCall, annotations: AnnotationHolderImpl) {
+        for (ann in annotations) {
+            mapAndCollectAnnotation(macro, ann)
+        }
+    }
+
+    private fun mapAndCollectAnnotation(macro: PreparedMacroCall, ann: Annotation) {
+        val originRange = TextRange(ann.startOffset, ann.endOffset)
+        val originInfo = HighlightInfo.fromAnnotation(ann)
+        val mappedRanges = mapRangeFromExpansionToCallBody(macro.expansion, macro.macroCall, originRange)
+        for (mappedRange in mappedRanges) {
+            results += originInfo.copyWithRange(mappedRange)
```

Comment:
Can we check here that there are two ranges with different attributes for the same offset? I think we should exclude them to avoid weird highlighting when the same token is annotated with different colors

Reply:
This requires complex data structures to be solved (or I need more time to think about it), to let's fox it later. I'll open an issue about it